### PR TITLE
using C_API of PATHSolver 1.1

### PR DESCRIPTION
**Pull request by chkwon**
Work in progress.

Issues
-  linear model indication
-  variable_names, constraint_names

### CODE REVIEW

File: MCP.md [hunk 1]
```diff
@@ -84,19 +82,18 @@ This is to define expressions for `F` in MCP. This is merely an alias of `JuMP.@
...
+solveMCP(m; convergence_tolerance=1e-8, output="yes", time_limit=3600)
 ```
-This solves the MCP and stores the solution inside `m`, which can be accessed by `getvalue(x)` as in JuMP.
+This solves the MCP and stores the solution inside `m`, which can be accessed by `result_value(x)`.
+Keyword arguments are options of the PATH Solver. See the [list of options](http://www.cs.wisc.edu/~ferris/path/options.pdf).
 
 
 ## Linear Complementarity Problems
 
+**NOT Supported with the new PATHSolver**
```
Comment (odow) [inline MCP.md]: Why have a special `linear = true`?

File: src/mcp.jl [hunk 2]
```diff
@@ -84,22 +84,33 @@ function sortperm_MCP_data(obj::Array{ComplementarityType,1})
 end
 
 # Using PATHSolver
-function _solve_path(m::JuMP.Model; linear=false)
+function _solve_path(m::JuMP.Model; linear=false, kwargs...)
```
Comment (odow) [inline src/mcp.jl]: We should lift the `NLPEvaluator` out of `myfunc` and `myjac`. 

```Julia
function _solve_path(m::JuMP.Model; linear = false, kwargs...)
    d = JuMP.NLPEvaluator(m)
    MOI.initialize(d, [:Jac])  # Shouldn't this be Jac not :Grad?
    J_struct = MOI.jacobian_structure(d)
    nnz = length(J_struct)
    I = first.(J_struct)
    J = last.(J_struct)
```

File: src/mcp.jl [hunk 3]
```diff
@@ -142,17 +168,25 @@ function _solve_path(m::JuMP.Model; linear=false)
...
+    # end
 
-        if !isapprox(J0, Jr)
-            error("The mappings do not seem linear. Rerun 'solveMCP()' after removing 'linear=true'.")
-        end
-        status, z, f = PATHSolver.solveLCP(myfunc, J0, lb, ub, initial_values, var_name, F_name)
-    else
-        status, z, f = PATHSolver.solveMCP(myfunc, myjac, lb, ub, initial_values, var_name, F_name)
-    end
+    status, z, info = PATHSolver.solve_mcp(myfunc, myjac, lb, ub, initial_values; kwargs...)
```
Comment (odow) [inline src/mcp.jl]: No need for `linear = true`? This should work even if the function is linear?

Comment (odow) [reply src/mcp.jl]: We could pass `nnz = nnz` here to reduce the size of the vectors created.

File: MCP.md (hunk 1 above)
Comment (chkwon) [reply MCP.md]: Sometime ago, people wanted to make PATHSolver.jl speedy for LCP by not calculating Jacobian multiple times. So it had `solveLCP` in addition to `solveMCP`, and `linear=true` is in the same line.

Comment (odow) [reply MCP.md]: I'd just remove it until people complain again. If people are solving linear problems and care that much, they can use `PATHSolver` directly.

File: src/mcp.jl (hunk 2 above)
Comment (chkwon) [reply src/mcp.jl]: Can `MOI.initialize(d, [:Jac])` be also lift out?

File: MCP.md (hunk 1 above)
Comment (chkwon) [reply MCP.md]: Sounds fair.

File: src/mcp.jl [hunk 4]
```diff
@@ -61,17 +61,14 @@ function get_initial_values_in_raw_index(m, mcp_data)
     return initial_values
 end
 
-function solveMCP(m::JuMP.Model; solver=:PATH, method=:trust_region, linear=false)
+function solveMCP!(m::JuMP.Model; solver=:PATH, method=:trust_region, kwargs...)
```
Comment (odow) [inline src/mcp.jl]: Any reason why renaming? This just breaks things. I would still take the `linear=false` keyword, but just do nothing with it.

```julia
function solveMCP(m::JuMP.Model; solver=:PATH, method=:trust_region, linear = false, kwargs...)
    if linear
        @warn("The linear keyword argument has been deprecated. You don't need to set it anymore.")
    end
```

File: examples/mcp/transmcp.jl [hunk 5]
```diff
@@ -55,17 +55,15 @@ using Test
...
-    status = solveMCP(m)
+    status = solveMCP!(m; convergence_tolerance=1e-8, output="yes", time_limit=3600)
 
     @show result_value.(x)
     @show result_value.(w)
     @show result_value.(p)
 
     @show status
-    @test status == :Solved
+    @test status == PATHSolver.MCP_Solved
```
Comment (odow) [inline examples/mcp/transmcp.jl]: Changing the return status is also breaking. Is there a need to?

File: src/mcp.jl (hunk 4 above)
Comment (chkwon) [reply src/mcp.jl]: Agreed. Thanks!

File: examples/mcp/transmcp.jl (hunk 5 above)
Comment (chkwon) [reply examples/mcp/transmcp.jl]: You are right. No good reason to break existing users' codes.

File: MCP.md (hunk 1 above)
Comment (davidanthoff) [reply MCP.md]: Also, if I understand things correctly, for linear problems one can now use JuMP directly, right? So for linear problems Complementarity.jl isn't really needed?

Comment (chkwon) [reply MCP.md]: @davidanthoff Correct.

**Pull request merged**